engine/gamesys/src/gamesys/components/comp_mesh.cpp

```diff
@@ -45,46 +46,114 @@ namespace dmGameSystem
         void*                              m_WorldVertexData;
         size_t                             m_WorldVertexDataSize;
         /// Keep track of how much vertex data we have rendered so it can be
         /// reported to the profiler at end of the render dispatching.
         uint32_t                           m_RenderedVertexSize;
 
-        dmObjectPool<MeshComponent*>       m_Components;
-        dmArray<dmRender::RenderObject>    m_RenderObjects;
     };
 
+    static inline void DeallocVertexBuffer(MeshWorld* world, dmGraphics::HVertexBuffer vertex_buffer)
+    {
+        if (world->m_VertexBufferPool.Full())
+            world->m_VertexBufferPool.OffsetCapacity(4);
+        world->m_VertexBufferPool.Push(vertex_buffer);
+    }
+
+    static inline dmGraphics::HVertexBuffer AllocVertexBuffer(MeshWorld* world, dmRender::HRenderContext render_context)
+    {
+        if (!world->m_VertexBufferPool.Empty())
+        {
+            dmGraphics::HVertexBuffer vertex_buffer = world->m_VertexBufferPool[world->m_VertexBufferPool.Size()-1];
+            world->m_VertexBufferPool.SetSize(world->m_VertexBufferPool.Size()-1);
+            return vertex_buffer;
+        }
+        dmGraphics::HContext graphics_context = dmRender::GetGraphicsContext(render_context);
+        // TODO: perhaps control stream vs draw better?
+        return dmGraphics::NewVertexBuffer(graphics_context, 0, 0x0, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+    }
+
+    static dmGraphics::HVertexBuffer GetVertexBuffer(MeshWorld* world, dmhash_t path)
+    {
+        VertexBufferInfo* info = world->m_ResourceToVertexBuffer.Get(path);
+        if (!info)
+            return 0;
+        return info->m_VertexBuffer;
+    }
+
+    static void AddVertexBuffer(MeshWorld* world, dmhash_t path, dmGraphics::HVertexBuffer vertex_buffer, uint32_t version)
+    {
+        VertexBufferInfo info;
+        info.m_RefCount = 1;
+        info.m_VertexBuffer = vertex_buffer;
+        info.m_Version = version;
+        if (world->m_ResourceToVertexBuffer.Full()) {
+            uint32_t capacity = world->m_ResourceToVertexBuffer.Capacity() + 8;
+            world->m_ResourceToVertexBuffer.SetCapacity(capacity/3, capacity);
+        }
+        world->m_ResourceToVertexBuffer.Put(path, info);
+    }
+
+    static void DecRefVertexBuffer(MeshWorld* world, dmhash_t path)
+    {
+        VertexBufferInfo* info = world->m_ResourceToVertexBuffer.Get(path);
+        assert(info != 0);
+        info->m_RefCount--;
+        if (info->m_RefCount == 0) {
+            world->m_ResourceToVertexBuffer.Erase(path);
+
+            DeallocVertexBuffer(world, info->m_VertexBuffer);
+        }
+    }
+
+    static void IncRefVertexBuffer(MeshWorld* world, dmhash_t path)
+    {
+        VertexBufferInfo* info = world->m_ResourceToVertexBuffer.Get(path);
+        assert(info != 0);
+        info->m_RefCount++;
+    }
```

Comment:
In order to keep track of these resources (vertex buffers), we reference count them.